crates/core/tedge_api/goals.md

```diff
@@ -0,0 +1,363 @@
+
+## Starting Thin-Edge
+
+On startup `tedge` will:
+
+- Check if your configuration is syntactically correct and that all requested
+  kinds exist
+- Startup the requested plugins and process messages
+
+In this case, the `heartbeat` service will keep sending messages every 400ms to
+both `watch_sudo_calls` and `check_service_alive` with a
+`MessageKind::SignalPluginState` to which they should answer with their
+status, e.g. `PluginStatus::Ok`.
+
+## Writing your own plugins
+
+### For External Plugins
+
+External plugins are executables that interact with the `tedge` api through a
+specific compile-time plugin, like for example `StdIoExternalPlugin` which
+communicates through STDIN/STDOUT.
+
+To use it, simply choose the plugin that fits your communication style and move
+to the configuration section below.
+
+### For Compile Plugins
+
+Compile-time plugins are written in Rust. They require two parts: A
+`PluginBuilder` and a `Plugin`. 
+
+- The `PluginBuilder` creates instances of `Plugins` and verifies if the
+  configuration is sane.
+- The `Plugin` is the object that contains the 'business logic' of your plugin
+
+So, to get started:
+
+1. Implement `PluginBuilder` for the struct that you will use to instantiate your plugin.
+
+It looks like this:
+
+```rust
+/// A plugin builder for a given plugin
+#[async_trait]
+pub trait PluginBuilder: Sync + Send + 'static {
+    /// The a name for the kind of plugins this creates, this should be unique and will prevent startup otherwise
+    fn kind_name(&self) -> &'static str;
+
+    /// This may be called anytime to verify whether a plugin could be instantiated with the
+    /// passed configuration.
+    async fn verify_configuration(&self, config: &PluginConfiguration) -> Result<(), PluginError>;
+
+    /// Instantiate a new instance of this plugin using the given configuration
+    ///
+    /// This _must not_ block
+    async fn instantiate(
+        &self,
+        config: PluginConfiguration,
+        tedge_comms: Comms,
+    ) -> Result<Box<dyn Plugin + 'static>, PluginError>;
+}
+```
+
+Things of note here:
+
+- `kind_name` _has_ to be unique, and will be used to refer to this kind of plugin
+- `verify_configuration` allows one to check if a given configuration _could even work_
+    - It however is not required to prove it
+- `instantiate` which actually constructs your plugin and returns a new instance of it
+    - One argument is the `Comms` object which holds the sender part of a
+      channel to the tedge core, and through which messages can be sent
+
+
+2. Implement `Plugin` for your plugin struct.
+
+`Plugin` is defined as follows:
+
+```rust
+/// A functionality extension to ThinEdge
+#[async_trait]
+pub trait Plugin: Sync + Send {
+    /// The plugin can set itself up here
+    async fn setup(&mut self) -> Result<(), PluginError>;
+
+    /// Handle a message specific to this plugin
+    async fn handle_message(&self, message: Message) -> Result<(), PluginError>;
```

Comment:
What is the API to send messages to other plugins?

Reply:
There's the `Comms` type ([see here](https://github.com/thin-edge/thin-edge.io/pull/948/files#diff-5babb1d99a4ee3af85937fe733b562e871310499187ba6fd352015996f9ad180R19-R72)) for that.